### PR TITLE
Fix threshold_triangle to work with non-integer images.

### LINKED ISSUE #2170: threshold_triangle fails for non-integer images
## Description
Thresholding with the triangle method can fail or give unreliable results for non-integer images. This is because of how the histogram is constructed, and because downstream of that there are a few places where bin number is equated with image data values. Observed when trying to use skimage.filters.try_all_threshold(). 

I'd be happy to make a PR with a fix.

## Way to reproduce
```python
from skimage.filters import thresholding

image = skimage.data.text()

# expected behavior, from the unit test
thresholding.threshold_triangle(image)
104

# by default uses 256 bins
thresholding.threshold_triangle(image.astype(float))
191.365234375

# should bin the same as when it's an integer image
thresholding.threshold_triangle(image.astype(float), nbins=image.max())
147.474619289

# expected behavior, from the unit test
thresholding.threshold_triangle(np.invert(image))
151

# bin centers are not integers but are used as indices
thresholding.threshold_triangle(np.invert(image).astype(float))
IndexError                                Traceback (most recent call last)
<ipython-input-110-04d94652cd41> in <module>()
----> 1 thresholding.threshold_triangle(np.invert(image).astype(float))

/python2.7/site-packages/skimage/filters/thresholding.pyc in threshold_triangle(image, nbins)
    720     width = arg_peak_height - arg_low_level
    721     x1 = np.arange(width)
--> 722     y1 = hist[x1 + arg_low_level]
    723 
    724     # Normalize.

IndexError: arrays used as indices must be of integer (or boolean) type
```

[If reporting a bug, please include the following important information:]
- [x] Code example
- [ ] Relevant images (if any)
- [ ] Operating system and version
- [x] Python version
Python 2.7.11
- [x] scikit-image version
0.13dev

### PR DESCRIPTION
**Pull request by rfinck**
## Description
This resolves #2170.

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
Resolves #2170.

### CODE REVIEW

File: skimage/filters/tests/test_thresholding.py [hunk 1]
```diff
@@ -321,8 +321,21 @@ def test_mean():
 
 
 def test_triangle_images():
-    assert(threshold_triangle(np.invert(data.text())) == 151)
-    assert(threshold_triangle(data.text()) == 104)
+    # Threshold uint8 image.
+    text = data.text()
+    assert(threshold_triangle(text) == 104)
+    int_bins = text.max() - text.min() + 1
+    # Threshold same image as float, setting nbins to match the integer case.
```
Comment (sciunto) [inline skimage/filters/tests/test_thresholding.py]: Perhaps, we can split this test. Keep one for uint and one for float.

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -727,14 +727,10 @@ def threshold_triangle(image, nbins=256):
     width /= norm
 
     # Maximize the length.
-    d = peak_height * arg_low_level - width * hist[arg_low_level]
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Why did you remove this line?

Comment (rfinck) [inline skimage/filters/thresholding.py]: This is a constant across the x-coordinates of the triangle, so it has no effect on minimizing the length, i.e.,  d/dx f(x) + c  = d/dx f(x). 

To be honest, I was able to reproduce the equation for how the length varies over the triangle, but rather than making sure that that shift was the correct offset to measure the absolute length, when I realized it didn't affect the minimum I simply removed it.

File: skimage/filters/tests/test_thresholding.py (hunk 1 above)
Comment (rfinck) [inline skimage/filters/tests/test_thresholding.py]: Can do. Note that I'm taking the 'correct' thresholds for these sample images from the original test as ground truth, so a float test is really asking 'does it get the equivalent value as the uint test'?

File: skimage/filters/thresholding.py (hunk 2 above)
Comment (sciunto) [inline skimage/filters/thresholding.py]: got it. However, I would prefer to keep a comment on that. Would you like to add that?

Comment (sciunto) [inline skimage/filters/thresholding.py]: It's just for code readers.

**Comment by sciunto** [comment]
I just left a last comment. After that, i'm OK to merge this PR.

Thanks @rfinck !

File: skimage/filters/thresholding.py (hunk 2 above)
Comment (rfinck) [inline skimage/filters/thresholding.py]: I'm all for additional comments there to explain the calculation -- I had to look at the reference to see the geometry -- but I'm still not sure I understand where that d came from. I got the equation that is now line 730 from comparing some similar triangles. Do you have a suggested comment that explains the original equation?

Comment (sciunto) [inline skimage/filters/thresholding.py]: I need to have a look too. I actually reused the implementation in imagej. I also read the paper, and the description in there is pretty short.

Comment (rfinck) [inline skimage/filters/thresholding.py]: Then I would suggest adding ImageJ as reference and adding a comment with something like 'the ImageJ implementation includes an additional constant when calculating the length, but here we omit it as it does not affect the location of the minimum'.

Also, I've attached my understanding of the trigonometry. 
img_20160628_085735

Comment (sciunto) [inline skimage/filters/thresholding.py]: I come up to the same result. I'm +1 to add the sentence you suggest as a comment

File: skimage/filters/thresholding.py [hunk 3]
```diff
@@ -687,6 +687,7 @@ def threshold_triangle(image, nbins=256):
     .. [1] Zack, G. W., Rogers, W. E. and Latt, S. A., 1977,
        Automatic Measurement of Sister Chromatid Exchange Frequency,
        Journal of Histochemistry and Cytochemistry 25 (7), pp. 741-753
+    .. [2] ImageJ AutoThresholder code, http://fiji.sc/wiki/index.php/Auto_Threshold
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: There is a warning treated as an error here.

**Comment by sciunto** [comment]
@rfinck Travis failed. Can you check please?

**Comment by rfinck** [comment]
Looks ok to me? 

Thanks for the review -- it's great to see the thresholding options being added.

**Comment by sciunto** [comment]
Perfect. LGTM.

PS: if you know how to squash the commits, it's even better.

@soupault Would you like to make the second review? It's easy :)

File: skimage/filters/thresholding.py (hunk 3 above)
Comment (soupault) [inline skimage/filters/thresholding.py]: This is 84 chars long. Move URI to the next line.

File: skimage/filters/tests/test_thresholding.py [hunk 4]
```diff
@@ -320,13 +320,28 @@ def test_mean():
...
+    int_bins = text.max() - text.min() + 1
+    # Set nbins to match the uint case and threshold as float.
+    assert(round(threshold_triangle(
+        text.astype(np.float), nbins=int_bins)) == 104)
+    # Check that rescaling image to floats in unit interval is equivalent.
+    assert(round(threshold_triangle(text / 255., nbins=int_bins) * 255) == 104)
+    # Repeat for inverted image.
+    assert(round(threshold_triangle(
+        np.invert(text).astype(np.float), nbins=int_bins)) == 151)
+    assert (round(threshold_triangle(
```
Comment (soupault) [inline skimage/filters/tests/test_thresholding.py]: That whitespace though :grimacing:.

The indentation here is weird, but, I support, it cannot be done better.

**Comment by soupault** [comment]
@rfinck just a one style comment. Once addressed, this goes in. :)

**Comment by rfinck** [comment]
Thanks @soupault.

**Comment by soupault** [comment]
@rfinck you too! :)
